### PR TITLE
[SPARK-16566][MLLib] sort sparseVector's indices before doing multiplication

### PR DESCRIPTION
**Pull request by wilson-lauw**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-16566

sort sparseVector's indices before doing multiplication to make sure the result returned correctly
## How was this patch tested?

manual and existing tests

### CODE REVIEW

**Comment by hhbyyh** [comment]
Order of the indices in SparseVector potentially affects many other BLAS operations. Perhaps we should have it in a more general/reusable pattern if we do want to impose the validation. 

Yet the most efficient way as I see is still to perform the check during construction, maybe something like an optional `validate` parameter.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala [hunk 1]
```diff
@@ -617,8 +617,27 @@ private[spark] object BLAS extends Serializable with Logging {
       x: SparseVector,
       beta: Double,
       y: DenseVector): Unit = {
-    val xValues = x.values
-    val xIndices = x.indices
+    
+    def isSorted(l:Array[Int]): Boolean = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala]: Many small style issues here. I think this is best as 

```
private def isSorted(array: Array[Int]): Boolean = {
  var index = 1
  while (index < array.length) {
    if (array(index - 1) > array(index)) {
      return false
    }
    index += 1
  }
  true
}
```

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala [hunk 2]
```diff
@@ -617,8 +617,27 @@ private[spark] object BLAS extends Serializable with Logging {
...
+        val prevElement = l(index - 1)
+        val currentElement = l(index)
+        
+        if(prevElement > currentElement) return false
+        index += 1
+      }
+      return true
+    }
+    
+    val (xValues,xIndices) = isSorted(x.indices) match{
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala]: match is overkilll for a boolean. Just use an `if`. We should avoid the style that omits member dereference "." here. Also you can use `unzip` I think

```
val (xValues, xIndices) =
  if (isSorted(x.indices)) {
    (x.values, x.indices)
  } else {
    x.values.zip(x.indices).sortBy(_._2).unzip
  }
```

The `def isSorted` probably should just be inlined, even, or made non-local

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala (hunk 1 above)
Comment (wilson-lauw) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala]: Thanks for the suggestion, will fix it.

**Comment by wilson-lauw** [comment]
file org.apache.spark.ml.linalg.BLAS.scala on module mllib-local also need to be fixed, once the changes is okay, I will also apply the changes there.

**Comment by wilson-lauw** [comment]
@hhbyyh I just checked, the implementation of `def dot(x: SparseVector, y: SparseVector)` also need to be modified.

**Comment by wilson-lauw** [comment]
Modifications also needed on module mllib-local, file org.apache.spark.ml.linalg.Vectors.scala. Once the changes is okay, I will also apply the changes there.

**Comment by jkbradley** [comment]
SparseVector indices are assumed to be sorted already.  We should not re-sort them during operations.  The deeper issue which needs to be fixed is SPARK-14707
I commented in the JIRA as well.
Could you please close this issue?
Thanks!

**Comment by wilson-lauw** [comment]
Closing this.

**Pull request closed without merging**